network/peer-monitoring-service/client/src/peer_states/mod.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{metrics, Error, PeerMonitorState, PeerMonitoringServiceClient, PeerState};
+use aptos_config::network_id::PeerNetworkId;
+use aptos_network::application::{interface::NetworkClient, metadata::PeerMetadata};
+use aptos_peer_monitoring_service_types::PeerMonitoringServiceMessage;
+use aptos_time_service::TimeService;
+use key_value::PeerStateKey;
+use std::collections::HashMap;
+use tokio::runtime::Handle;
+
+mod key_value;
+mod latency_info;
+mod network_info;
+pub mod peer_state;
+mod request_tracker;
+
+/// Refreshes the states of the connected peers
+pub fn refresh_peer_states(
+    peer_monitor_state: PeerMonitorState,
+    peer_monitoring_client: PeerMonitoringServiceClient<
+        NetworkClient<PeerMonitoringServiceMessage>,
+    >,
+    connected_peers_and_metadata: HashMap<PeerNetworkId, PeerMetadata>,
+    time_service: TimeService,
+    runtime: Option<Handle>,
+) -> Result<(), Error> {
+    // Process all state entries (in order) and update the ones that
+    // need to be refreshed for each peer.
+    for peer_state_key in PeerStateKey::get_all_keys() {
+        let mut num_in_flight_requests = 0;
+
+        // Go through all connected peers and see if we should refresh the state
+        for (peer_network_id, peer_metadata) in &connected_peers_and_metadata {
+            // Get the peer state
+            let peer_state = get_peer_state(&peer_monitor_state, peer_network_id)?;
+
+            // If there's an-flight request, update the metrics counter
+            let request_tracker = peer_state.get_request_tracker(&peer_state_key)?;
+            if request_tracker.read().in_flight_request() {
+                num_in_flight_requests += 1;
+            }
+
+            // Update the state if it needs to be refreshed
+            let should_refresh_peer_state_key = request_tracker.read().new_request_required();
```

Comment:
not sure if it matters, but:
- do we want to send pings if we had other messages exchanged recently, that we can piggyback on?
- do we want to create a large burst of pings at the same time, or add some small delay between each (i.e. at least 10ms?

Reply:
> do we want to send pings if we had other messages exchanged recently, that we can piggyback on?

In the ideal world, yes 😄 But, there's probably too much tech debt to do this immediately. Also, the benefit of a small dedicated message is that we get reasonable ping latencies (as opposed to piggy backing on larger messages that might increase the message time because of data sizes)? I think we can probably explore this in the future, but, I'm not super worried for now.

> do we want to create a large burst of pings at the same time, or add some small delay between each (i.e. at least 10ms?

Good question. I think there should be enough randomness here already because the timer starts at the moment of the peer to peer connection, which already has some small randomness. But, I can also follow up with a very simple mechanism to add some jitter :)